# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGL2ContextBuffers.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGL2ContextBuffers.cpp
--- a/dom/canvas/WebGL2ContextBuffers.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGL2ContextBuffers.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -5,6 +5,7 @@
 
 #include "WebGL2Context.h"
 
+#include "ClientWebGLContext.h"
 #include "GLContext.h"
 #include "WebGLBuffer.h"
 #include "WebGLTransformFeedback.h"
@@ -84,35 +85,22 @@
   writeBuffer->ResetLastUpdateFenceId();
 }
 
-void WebGL2Context::GetBufferSubData(GLenum target, WebGLintptr srcByteOffset,
-                                     const dom::ArrayBufferView& dstData,
-                                     GLuint dstElemOffset,
-                                     GLuint dstElemCountOverride) {
+Maybe<UniquePtr<RawBuffer<>>> WebGL2Context::GetBufferSubData(
+    GLenum target, WebGLintptr srcByteOffset, size_t byteLen) {
   const FuncScope funcScope(*this, "getBufferSubData");
-  if (IsContextLost()) return;
-
-  if (!ValidateNonNegative("srcByteOffset", srcByteOffset)) return;
-
-  uint8_t* bytes;
-  size_t byteLen;
-  if (!ValidateArrayBufferView(dstData, dstElemOffset, dstElemCountOverride,
-                               LOCAL_GL_INVALID_VALUE, &bytes, &byteLen)) {
-    return;
-  }
-
-  ////
+  if (IsContextLost()) return Nothing();
 
   const auto& buffer = ValidateBufferSelection(target);
-  if (!buffer) return;
+  if (!buffer) return Nothing();
 
-  if (!buffer->ValidateRange(srcByteOffset, byteLen)) return;
+  if (!buffer->ValidateRange(srcByteOffset, byteLen)) return Nothing();
 
   ////
 
   if (!CheckedInt<GLintptr>(srcByteOffset).isValid() ||
       !CheckedInt<GLsizeiptr>(byteLen).isValid()) {
     ErrorOutOfMemory("offset or size too large for platform.");
-    return;
+    return Nothing();
   }
   const GLsizeiptr glByteLen(byteLen);
 
@@ -138,6 +126,12 @@
 
   ////
 
+  uint8_t* bytes = new uint8_t[byteLen];
+  if (!bytes) {
+    ErrorOutOfMemory("Could not allocate temp array");
+    return Nothing();
+  }
+
   const ScopedLazyBind readBind(gl, target, buffer);
 
   if (byteLen) {
@@ -162,6 +156,9 @@
       gl->fBindTransformFeedback(LOCAL_GL_TRANSFORM_FEEDBACK, tfo);
     }
   }
+
+  UniquePtr<RawBuffer<>> ret = MakeUnique<RawBuffer<>>(byteLen, bytes, true);
+  return Some(std::move(ret));
 }
 
 }  // namespace mozilla